skimage/draw/draw.py

```diff
@@ -693,3 +693,77 @@ def bezier_curve(r0, c0, r1, c1, r2, c2, weight, shape=None):
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
     """
     return _bezier_curve(r0, c0, r1, c1, r2, c2, weight, shape)
+
+
+def rectangle(start, end = None, extent = None, shape=None):
+    """Generate coordinates of pixels within an nd-rectangle. Either
+    end or extent are required but not both: (extent is None) ^ (end is None)
+
+    Parameters
+    ----------
+    start : tuple
+        origin point of the rectangle, ([plane], row, column[, ...])
+    end : tuple
+        end point of the rectangle ([num_planes], num_rows, num_cols[, ...]).
+    extent : tuple
+        extent vector ([num_planes], num_rows, num_cols[, ...]) of the
+        rectangle. Will be added to start to find the end point.
+    shape : tuple, optional
+        Image shape which is used to determine the maximum bounds of the output
+        pixel coordinates. This is useful for rectangles that exceed the image
+        size. If None, the bounds are determined from the bounds of
+        the rectangle.
+
+    Returns
+    -------
+    coords : array of int, shape (Ndim, Npoints)
+        The coordinates of all pixels in the nD rectangle.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skimage.draw import rectangle
+    >>> img = np.zeros((5, 5), dtype=np.uint8)
+    >>> start = (1, 1)
+    >>> extent = (3, 3)
+    >>> rr, cc = rectangle(start, extent = extent, shape = img.shape)
+    >>> img[rr, cc] = 1
+    >>> img
+    array([[0, 0, 0, 0, 0],
+           [0, 1, 1, 1, 0],
+           [0, 1, 1, 1, 0],
+           [0, 1, 1, 1, 0],
+           [0, 0, 0, 0, 0]], dtype=uint8)
+
+
+    >>> img = np.zeros((5, 5), dtype=np.uint8)
+    >>> start = (0, 1)
+    >>> end = (3, 3)
+    >>> rr, cc = rectangle(start, end = end, shape = img.shape)
+    >>> img[rr, cc] = 1
+    >>> img
+    array([[0, 1, 1, 1, 0],
+           [0, 1, 1, 1, 0],
+           [0, 1, 1, 1, 0],
+           [0, 1, 1, 1, 0],
+           [0, 0, 0, 0, 0]], dtype=uint8)
+
+    """
+    if end is None:
+        if extent is None:
+            raise ValueError("either an end or extent must be given")
+        else:
+            end = np.array(start) + np.array(extent)
+    else:
+        if extent is not None:
+            raise ValueError("only end or extent can be given not both")
+    bl = np.minimum(np.array(start), np.array(end))
```

Comment:
I'm assuming these mean "bottom left" and "top right". However, since we are using NumPy coordinates, I think top left and bottom right are more appropriate. Also, you don't need to convert to `np.array` manually: almost all NumPy functions will do this for you. So I suggest `np.minimum(start, end)`. Same for the other `np.minimum/maximum` calls below.